File: src/main/java/com/uiptv/model/Account.java
package com.uiptv.model;


import com.uiptv.shared.BaseJson;
import com.uiptv.util.AccountType;
import com.uiptv.util.StringUtils;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import org.json.JSONPropertyIgnore;

import java.util.Arrays;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import static com.uiptv.model.Account.AccountAction.itv;
import static com.uiptv.util.AccountType.*;
import static com.uiptv.util.StringUtils.SPACE;
import static com.uiptv.util.StringUtils.isNotBlank;

@Data
@NoArgsConstructor()
@EqualsAndHashCode(callSuper = false)
public class Account extends BaseJson {

    @SuppressWarnings("java:S115")
    public enum AccountAction {
        itv,
        vod,
        series
    }

    public static final Set<AccountAction> NOT_LIVE_TV_CHANNELS = Collections.unmodifiableSet(EnumSet.of(AccountAction.vod, AccountAction.series));
    public static final Set<AccountType> VOD_AND_SERIES_SUPPORTED = Collections.unmodifiableSet(EnumSet.of(STALKER_PORTAL, XTREME_API));
    public static final Set<AccountType> CACHE_SUPPORTED = Collections.unmodifiableSet(EnumSet.of(STALKER_PORTAL, XTREME_API, M3U8_URL, M3U8_LOCAL));
    public static final Set<AccountType> PRE_DEFINED_URLS = Collections.unmodifiableSet(EnumSet.of(RSS_FEED, M3U8_URL, M3U8_LOCAL));
    public static final String LINE_SEPARATOR = "\n\r";
    private String serverPortalUrl;
    private AccountAction action = itv;
    private String accountName;
    private String username;
    private String password;
    private String xtremeCredentialsJson;
    private String url;
    private String macAddress;
    private String macAddressList;
    private String serialNumber;
    private String deviceId1;
    private String deviceId2;
    private String signature;
    private String epg;
    private String m3u8Path;
    private String dbId;
    private String token;
    private boolean pinToTop;
    private boolean resolveChainAndDeepRedirects;
    private AccountType type = STALKER_PORTAL;
    private String httpMethod = "GET";
    private String timezone = "Europe/London";

    @SuppressWarnings("java:S107")
    public Account(String accountName, String username, String password, String url, String macAddress, String macAddressList, String serialNumber, String deviceId1, String deviceId2, String signature, AccountType type, String epg, String m3u8Path, boolean pinToTop) {
        this.accountName = accountName;
        this.username = username;
        this.password = password;
        this.url = url;
        this.macAddress = macAddress;
        this.serialNumber = serialNumber;
        this.deviceId1 = deviceId1;
        this.deviceId2 = deviceId2;
        this.signature = signature;
        this.type = type;
        this.epg = epg;
        this.m3u8Path = m3u8Path;
        Map<String, String> macMap = new HashMap<>();
        if (isNotBlank(macAddress)) macMap.put(macAddress.toLowerCase(), macAddress.trim());
        if (isNotBlank(macAddressList)) {
            Arrays.stream(macAddressList.split(","))
                    .filter(StringUtils::isNotBlank)
                    .forEach(m -> macMap.put(m.toLowerCase(), m.replace(SPACE, "")));
        }
        this.macAddressList = String.join(",", macMap.values());
        this.pinToTop = pinToTop;
    }

    @JSONPropertyIgnore
    public boolean isConnected() {
        if (type != STALKER_PORTAL) return true;
        return isNotBlank(token);
    }

    @JSONPropertyIgnore
    public boolean isNotConnected() {
        return !isConnected();
    }
}

File: server/src/main/kotlin/com/uiptv/model/Account.kt
package com.uiptv.model

import com.uiptv.shared.BaseJson
import com.uiptv.util.AccountType
import com.uiptv.util.StringUtils
import java.util.Arrays
import java.util.Collections
import java.util.EnumSet
import java.util.LinkedHashMap

import com.uiptv.model.Account.AccountAction.itv
import com.uiptv.util.AccountType.M3U8_LOCAL
import com.uiptv.util.AccountType.M3U8_URL
import com.uiptv.util.AccountType.RSS_FEED
import com.uiptv.util.AccountType.STALKER_PORTAL
import com.uiptv.util.AccountType.XTREME_API
import com.uiptv.util.StringUtils.SPACE
import com.uiptv.util.StringUtils.isNotBlank

data class Account @JvmOverloads constructor(
    var serverPortalUrl: String? = null,
    var action: AccountAction = itv,
    var accountName: String? = null,
    var username: String? = null,
    var password: String? = null,
    var xtremeCredentialsJson: String? = null,
    var url: String? = null,
    var macAddress: String? = null,
    var macAddressList: String? = null,
    var serialNumber: String? = null,
    var deviceId1: String? = null,
    var deviceId2: String? = null,
    var signature: String? = null,
    var epg: String? = null,
    var m3u8Path: String? = null,
    var dbId: String? = null,
    var token: String? = null,
    @get:JvmName("isPinToTop")
    var pinToTop: Boolean = false,
    @get:JvmName("isResolveChainAndDeepRedirects")
    var resolveChainAndDeepRedirects: Boolean = false,
    var type: AccountType = STALKER_PORTAL,
    var httpMethod: String = "GET",
    var timezone: String = "Europe/London"
) : BaseJson() {

    @Suppress("java:S107")
    constructor(
        accountName: String?,
        username: String?,
        password: String?,
        url: String?,
        macAddress: String?,
        macAddressList: String?,
        serialNumber: String?,
        deviceId1: String?,
        deviceId2: String?,
        signature: String?,
        type: AccountType,
        epg: String?,
        m3u8Path: String?,
        pinToTop: Boolean
    ) : this(
        accountName = accountName,
        username = username,
        password = password,
        url = url,
        macAddress = macAddress,
        serialNumber = serialNumber,
        deviceId1 = deviceId1,
        deviceId2 = deviceId2,
        signature = signature,
        type = type,
        epg = epg,
        m3u8Path = m3u8Path,
        pinToTop = pinToTop
    ) {
        val macMap = LinkedHashMap<String, String>()
        if (isNotBlank(macAddress)) {
            macMap[macAddress!!.lowercase()] = macAddress.trim()
        }
        if (isNotBlank(macAddressList)) {
            Arrays.stream(macAddressList!!.split(",").toTypedArray())
                .filter(StringUtils::isNotBlank)
                .forEach { mac -> macMap[mac.lowercase()] = mac.replace(SPACE, "") }
        }
        this.macAddressList = macMap.values.joinToString(",")
    }

    fun isConnected(): Boolean {
        if (type != STALKER_PORTAL) {
            return true
        }
        return isNotBlank(token)
    }

    fun isNotConnected(): Boolean = !isConnected()

    enum class AccountAction {
        itv,
        vod,
        series
    }

    companion object {
        @JvmField
        val NOT_LIVE_TV_CHANNELS = Collections.unmodifiableSet(EnumSet.of(AccountAction.vod, AccountAction.series))

        @JvmField
        val VOD_AND_SERIES_SUPPORTED = Collections.unmodifiableSet(EnumSet.of(STALKER_PORTAL, XTREME_API))

        @JvmField
        val CACHE_SUPPORTED = Collections.unmodifiableSet(EnumSet.of(STALKER_PORTAL, XTREME_API, M3U8_URL, M3U8_LOCAL))

        @JvmField
        val PRE_DEFINED_URLS = Collections.unmodifiableSet(EnumSet.of(RSS_FEED, M3U8_URL, M3U8_LOCAL))

        const val LINE_SEPARATOR: String = "\n\r"
    }
}
